artifact.go

```diff
@@ -203,67 +213,423 @@ func (m artifactModule) getBuildTestsResultsHandler(c *gin.Context) {
+// @tags artifact
+// @param buildid path int true "Build ID"
+// @param artifactId path int true "Artifact ID"
+// @success 200 {object} []TestResultDetail
+// @router /build/{buildid}/test-result-details/{artifactId} [get]
+func (m artifactModule) getBuildTestResultDetailsHandler(c *gin.Context) {
+	buildID, ok := ginutil.ParseParamUint(c, buildIDParamName)
+	if !ok {
+		return
+	}
+
+	artifactID, ok := ginutil.ParseParamUint(c, artifactIDParamName)
+	if !ok {
+		return
+	}
+
+	details := []TestResultDetail{}
+	err := m.Database.
+		Where(&TestResultDetail{BuildID: uint(buildID), ArtifactID: artifactID}).
+		Find(&details).
+		Error
+
+	if err != nil {
+		ginutil.WriteDBReadError(c, err, fmt.Sprintf(
+			"Failed fetching test result details from test with ID %d for build with ID %d from database.",
+			artifactID, buildID))
+		return
+	}
+
+	c.JSON(http.StatusOK, details)
+}
+
+// getBuildTestResultsSummaryHandler godoc
+// @summary Get test result summary of all tests for specified build
+// @tags artifact
+// @param buildid path int true "Build ID"
+// @success 200 {object} TestResultListSummary
+// @failure 400 {object} problem.Response "Bad Request"
+// @failure 502 {object} problem.Response "Bad Gateway"
+// @router /build/{buildid}/test-results-summary [get]
+func (m artifactModule) getBuildTestResultsSummaryHandler(c *gin.Context) {
+	buildID, ok := ginutil.ParseParamUint(c, buildIDParamName)
+	if !ok {
+		return
+	}
+
+	summaries := []TestResultSummary{}
+	err := m.Database.
+		Preload("Artifact").
+		Where(&TestResultSummary{BuildID: uint(buildID)}).
+		Find(&summaries).
+		Error
+
+	if err != nil {
+		ginutil.WriteDBReadError(c, err, fmt.Sprintf(
+			"Failed fetching test result summaries for build with ID %d from database.",
+			buildID))
+		return
+	}
+
+	summaryList := TestResultListSummary{
+		BuildID:   buildID,
+		Summaries: summaries}
+
+	for _, v := range summaries {
+		summaryList.Failed += v.Failed
+		summaryList.Passed += v.Passed
+		summaryList.Skipped += v.Skipped
+	}
+
+	summaryList.Total = summaryList.Failed + summaryList.Passed + summaryList.Skipped
+
+	c.JSON(http.StatusOK, summaryList)
+}
+
+func parseMultipartFormData(c *gin.Context, buildID uint) ([]file, bool) {
+	files := make([]file, 0)
+
+	form, err := c.MultipartForm()
+	if err != nil {
+		ginutil.WriteMultipartFormReadError(c, err,
+			fmt.Sprintf("Failed reading multipart-form content from request body when uploading new"+
+				" artifact for build with ID %d.", buildID))
+		return nil, false
+	}
+
 	for k := range form.File {
 		if fhs := form.File[k]; len(fhs) > 0 {
-			f, err := fhs[0].Open()
-			if err != nil {
-				ginutil.WriteMultipartFormReadError(c, err, fmt.Sprintf(
-					"Failed with starting to read file content from multipart form request body when uploading new artifact for build with ID %d.",
-					buildID))
-				return
+			data, ok := readMultipartFileData(c, buildID, fhs[0])
+			if !ok {
+				return files, false
```

Comment:
Vote fot returning nil here as well, as it failed halfway. So it's consistent with the returns for the other fail states, and that we don't expect to use the return value anyway.

```suggestion
				return nil, false
```